pkg/kds/mux/session.go

```diff
@@ -82,34 +99,62 @@ func (s *session) handleRecv(stream MultiplexStream) {
 
 // handleSend polls either sendBuffer and call send on the KDSStream (the actual grpc bidi-stream).
 // This call is stopped whenever either of the sendBuffer are closed (in practice they are always closed together anyway).
-func (s *session) handleSend(stream MultiplexStream) {
+func (s *session) handleSend(stream MultiplexStream, sendTimeout time.Duration) {
 	kdsVersion := KDSVersion(stream.Context())
 	for {
+		var msgToSend *mesh_proto.Message
 		select {
-		case item, more := <-s.serverStream.bufferStream.sendBuffer:
+		case msg, more := <-s.serverStream.bufferStream.sendBuffer:
 			if !more {
 				return
 			}
-			r := item.msg
 			if kdsVersion == KDSVersionV2 {
-				r = &mesh_proto.Message{Value: &mesh_proto.Message_LegacyResponse{LegacyResponse: DiscoveryResponseV2(r.GetResponse())}}
+				msg = &mesh_proto.Message{Value: &mesh_proto.Message_LegacyResponse{LegacyResponse: DiscoveryResponseV2(msg.GetResponse())}}
 			}
-			err := stream.Send(r)
-			item.errChan <- err
-		case item, more := <-s.clientStream.bufferStream.sendBuffer:
+			msgToSend = msg
+		case msg, more := <-s.clientStream.bufferStream.sendBuffer:
 			if !more {
 				return
 			}
-			r := item.msg
 			if kdsVersion == KDSVersionV2 {
-				r = &mesh_proto.Message{Value: &mesh_proto.Message_LegacyRequest{LegacyRequest: DiscoveryRequestV2(r.GetRequest())}}
+				msg = &mesh_proto.Message{Value: &mesh_proto.Message_LegacyRequest{LegacyRequest: DiscoveryRequestV2(msg.GetRequest())}}
 			}
-			err := stream.Send(r)
-			item.errChan <- err
+			msgToSend = msg
 		}
+		ctx, cancel := context.WithTimeout(context.Background(), sendTimeout)
+		go func() {
+			<-ctx.Done()
+			if ctx.Err() == context.DeadlineExceeded {
+				// This is very unlikely to happen, but it was introduced as a last resort protection from a gRPC streaming deadlock.
+				// gRPC streaming deadlock may happen if both peers are stuck on Send() operation without calling Recv() often enough.
+				// In this case, if data is big enough, both parties may wait for WINDOW_UPDATE on HTTP/2 stream.
+				// We fixed the deadlock by increasing buffer size which is larger that all possible inflight request.
+				// If the connection is broken and send is stuck, it's more likely for gRPC keep alive to catch such case.
+				// If you still hit the timeout without deadlock, you may increase it. However, there are two possible scenarios
+				// 1) This is a malicious client reading stream byte by byte. In this case it's actually better to end the stream
+				// 2) A client is such overwhelmed that it cannot even let the server know that it's ready to receive more data.
+				//    In this case it's recommended to scale number of instances.
+				s.setError(errors.New("timeout while sending a message to peer"))
+			}
+		}()
+		if err := stream.Send(msgToSend); err != nil {
```

Comment:
There is no way to set timeout on this operation therefore we need to use a separate goroutine.
The pattern of
```
check if timeout on background goroutine
call send on on main goroutine
```
instead of
```
call send on on background goroutine
wait for timeout on the main goroutine
```
was deliberately chosen to keep the unit tests that check that we don't call `Send` concurrently.